packages/codegen-ui-react/lib/action.ts

```diff
@@ -0,0 +1,137 @@
+  StudioComponentProperty,
+} from '@aws-amplify/codegen-ui';
+import {
+  isFixedPropertyWithValue,
+  isBoundProperty,
+  isConcatenatedProperty,
+  isConditionalProperty,
+  buildBindingExpression,
+  buildBindingWithDefaultExpression,
+  buildConcatExpression,
+  buildConditionalExpression,
+  buildFixedLiteralExpression,
+  isActionEvent,
+} from './react-component-render-helper';
+
+enum Action {
+  'Amplify.Navigate' = 'Amplify.Navigate',
+}
+
+export default Action;
+
+export const ActionNameMapping: Partial<Record<Action, string>> = {
+  [Action['Amplify.Navigate']]: 'useNavigateAction',
+};
+
+export function isAction(action: string): action is Action {
+  return Object.values(Action).includes(action as Action);
+}
+
+export function getActionName(action: string): string {
+  const actionName = ActionNameMapping[Action[action as Action]];
+  if (actionName === undefined) {
+    throw new InvalidInputError(`${action} is not a valid action.`);
+  }
+  return actionName;
+}
+
+export function getComponentActions(component: StudioComponent | StudioComponentChild): ActionStudioComponentEvent[] {
+  // TODO: add unique identifier to each action
+  if (component.events) {
+    const actions = Object.values(component.events).filter(isActionEvent);
+    if (component.children) {
+      return actions.concat(component.children.map(getComponentActions).flat());
+    }
+    return actions;
+  }
+  return [];
+}
+
+export function buildUseActionStatement(action: ActionStudioComponentEvent): Statement {
+  const actionName = getActionName(action.action);
+  return factory.createVariableStatement(
+    undefined,
+    factory.createVariableDeclarationList(
+      [
+        factory.createVariableDeclaration(
+          // TODO: update to unique name
+          factory.createIdentifier('action'),
+          undefined,
+          undefined,
+          factory.createCallExpression(factory.createIdentifier(actionName), undefined, [
+            buildActionParameters(action),
+          ]),
+        ),
+      ],
+      ts.NodeFlags.Const,
+    ),
+  );
+}
+
+/* Transform the action parameters field to literal
+ *
+ * model and fields are special cases. All other fields are StudioComponentProperty
+ */
+export function buildActionParameters(action: ActionStudioComponentEvent): ObjectLiteralExpression {
+  if (action.parameters) {
+    // TODO: add special case for model and fields
+    return factory.createObjectLiteralExpression(
+      Object.entries(action.parameters).map(
+        ([key, value]) =>
+          factory.createPropertyAssignment(
+            factory.createIdentifier(key),
+            actionParameterToExpression(value as StudioComponentProperty),
+          ),
+        false,
+      ),
+    );
+  }
+  // TODO: determine what should be used when no parameters
+  return factory.createObjectLiteralExpression([], false);
+}
+
+export function actionParameterToExpression(parameter: StudioComponentProperty): Expression {
```

Comment:
We almost certainly support a generic version of this already in the helper, right?

Reply:
Or are you keeping separate to make it easier to handle the circular dependency thing here?

Reply:
I think we actually don't have a generic version. I wrote this portion around 2 months ago so I can't remember the exact reasoning. It looks like there is repeated logic for isFixedProperty, isBoundProperty, etc. but each adds a different spin to the output. We should refactor all to use one generic version as the base.

* https://github.com/aws-amplify/amplify-codegen-ui/blob/main/packages/codegen-ui-react/lib/react-component-render-helper.ts#L227
* https://github.com/aws-amplify/amplify-codegen-ui/blob/main/packages/codegen-ui-react/lib/react-component-render-helper.ts#L264
* https://github.com/aws-amplify/amplify-codegen-ui/blob/main/packages/codegen-ui-react/lib/react-component-render-helper.ts#L370

I think it's best to do that as a separate change though.

Reply:
Yep, makes sense to me. I'll create a tracking task.